### PR TITLE
Homework1 done

### PR DESCRIPTION
**Pull request by iavicha**
Hi Artur!
Can u give me feedback about my code?

### CODE REVIEW

File: homework1/task02.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+We guarantee, that the given sequence contain >= 0 integers inside.
+
+"""
+from typing import Sequence
+
+
+def check_fibonacci(data: Sequence[int]) -> bool:
+
+    if len(data) == 0:
+        return False
```
Comment (apaniukov) [inline homework1/task02.py:14]: I would argue that if the sequence contains no elements it is a Fibonacci sequence. You just can't disprove it, can you?

File: homework1/tests/task02.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import pytest
+
+from homework1.task02 import check_fibonacci
+
+
+@pytest.mark.parametrize(
+    ["value", "expected_result"],
+    [([1, 1, 2, 3, 5, 8], True), ([1, 1, 2, 3, 5, 7], False)],
```
Comment (apaniukov) [inline homework1/tests/task02.py]: You should add more corner cases for this test: 0 elements in the list, 1 element, etc.

File: homework1/task03.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+with open("some_file.txt") as fi:
+    for line in fi:
+        ...
+
+"""
+from typing import Tuple
+
+
+def find_maximum_and_minimum(file_name: str) -> Tuple[int, int]:
+    result = ()
```
Comment (apaniukov) [inline homework1/task03.py]: Why you need a tuple outside the for-loop, if you constructed the new one with the answer inside the for-loop anyway?

File: homework1/task02.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    for index, numb in enumerate(data):
+        if index > 1:
+            if data[index - 2] + data[index - 1] != numb:
+                return False
+        elif index == 0 and numb not in (0, 1):
```
Comment (apaniukov) [inline homework1/task02.py]: Why not check the first two values outside the for-loop? Too many branches is bad for the modern processor architectures.

File: homework1/task03.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    with open(file_name, "r") as file:
+        for line in file:
+            file = list(line.replace(" ", "").split(","))
+            result = (int(min(file))), (int(max(file)))
+            return result
```
Comment (apaniukov) [inline homework1/task03.py]: You don't need to create a tuple separeatly, just `return min, max`

File: homework1/task03.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+"""
+from typing import Tuple
+
+
+def find_maximum_and_minimum(file_name: str) -> Tuple[int, int]:
+    result = ()
+
+    with open(file_name, "r") as file:
+        for line in file:
```
Comment (apaniukov) [inline homework1/task03.py:24]: This for loop is only executes one time.

File: homework1/tests/task03.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import pytest
+
+from homework1.task03 import find_maximum_and_minimum
+
+
+@pytest.mark.parametrize(
+    ["value", "expected_result"],
+    [
+        ("homework1/some_file.txt", (1, 5)),
```
Comment (apaniukov) [inline homework1/tests/task03.py]: If you add an 11 to the file, min and max will be 1 and 5. You shold add more test cases for that function.

File: homework1/task04.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+def check_sum_of_four(a: List[int], b: List[int], c: List[int], d: List[int]) -> int:
+    count = 0
+    for i in a:
+        for j in b:
+            for k in c:
+                for l in d:
+                    if i + j + k + l == 0:
+                        count += 1
+    return count
```
Comment (apaniukov) [inline homework1/task04.py:23]: Could you write a second solution, using an `itertools` library?

File: homework1/tests/task04.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import pytest
+
+from homework1.task04 import check_sum_of_four
+
+
+@pytest.mark.parametrize(
```
Comment (apaniukov) [inline homework1/tests/task04.py]: If you have only one test case, you shouldn't use `parametrize`, just hardcoded values in the test body. But I recomend to check more edge cases here.

File: homework1/tests/task05.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+from homework1.task05 import find_maximal_subarray_sum
+
+
+@pytest.mark.parametrize(
+    ["value", "value1", "expected_result"],
+    [
+        ([1, 3, -1, -3, 5, 3, 6, 7], 3, 16),
+    ],
+)
+def find_maximal_subarray_sum(value: list, value1: int, expected_result: int):
```
Comment (apaniukov) [inline homework1/tests/task05.py]: All test functions must start with `test_` prefix. Also you should name variables in a meaningful way, because `value` and  `value1` are hard to understand, and `input_list` and `k`(in the context of the task) is better.

File: homework1/tests/task05.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
+import pytest
+
+from homework1.task05 import find_maximal_subarray_sum
+
+
+@pytest.mark.parametrize(
+    ["value", "value1", "expected_result"],
+    [
+        ([1, 3, -1, -3, 5, 3, 6, 7], 3, 16),
```
Comment (apaniukov) [inline homework1/tests/task05.py]: ([-1, -2, -3], 3, -1) - wrong answer for the test case

**Review by apaniukov — CHANGES_REQUESTED** [review]
The test files must start with `test_` prefix.

File: homework1/tests/task05.py (hunk 10 above)
Comment (iavicha) [reply homework1/tests/task05.py]: fixed.

File: homework1/tests/task05.py (hunk 11 above)
Comment (iavicha) [reply homework1/tests/task05.py]: fixed.

File: homework1/tests/task02.py (hunk 2 above)
Comment (iavicha) [reply homework1/tests/task02.py]: fixed.

File: homework1/task03.py (hunk 3 above)
Comment (iavicha) [reply homework1/task03.py]: I don't need it.

File: homework1/task03.py (hunk 5 above)
Comment (iavicha) [reply homework1/task03.py]: I don't need it.

File: homework1/tests/task03.py (hunk 7 above)
Comment (iavicha) [reply homework1/tests/task03.py]: Fixed it.

File: homework1/task04.py (hunk 8 above)
Comment (iavicha) [reply homework1/task04.py:23]: featured it.

File: homework1/tests/task04.py (hunk 9 above)
Comment (iavicha) [reply homework1/tests/task04.py]: 👌

File: homework1/task02.py (hunk 4 above)
Comment (iavicha) [reply homework1/task02.py]: fixed.

File: homework1/task04.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+
+def check_sum_of_four_second_solution(
+    a: List[int], b: List[int], c: List[int], d: List[int]
+) -> int:
+    result = [i for i in itertools.product(a, b, c, d) if sum(i) == 0]
+    return len(result)
```
Comment (apaniukov) [inline homework1/task04.py:30]: Another option is just `return sum(1 for mumbers in itertools.product(a, b, c, d) if sum(numbers) == 0)

File: homework1/task05.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+def find_maximal_subarray_sum(nums: List[int], k: int) -> int:
+    if k <= 0:
+        exit("Think")
+
+    result = []
+    nums = deque(nums)
+
+    while len(nums) > 0:
+        for one in range(1, k + 1):
+            result.append(sum(list(itertools.islice(nums, one))))
```
Comment (apaniukov) [inline homework1/task05.py:26]: If you use `islice` inside `sum`, you don't have to construct `list`. It should work just fine without intermediate collection construction.

**Review by apaniukov — APPROVED** [review]
